### PR TITLE
[Uid] Fix validating nil and max uuid

### PR DESCRIPTION
**Pull request by fancyweb**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | https://github.com/symfony/symfony/issues/48800
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Component/Uid/Uuid.php [hunk 1]
```diff
@@ -117,6 +117,14 @@ final public static function v6(): UuidV6
 
     public static function isValid(string $uuid): bool
     {
+        if (self::NIL === $uuid && \in_array(static::class, [__CLASS__, NilUuid::class], true)) {
+            return true;
+        }
+
+        if (__CLASS__ === static::class && 'ffffffff-ffff-ffff-ffff-ffffffffffff' === strtr($uuid, 'F', 'f')) {
```
Comment (fancyweb) [inline src/Symfony/Component/Uid/Uuid.php:124]: `if (self::MAX === strtr($uuid, 'F', 'f')  && (\in_array(static::class, [__CLASS__, MaxUuid::class], true))) {` on 6.2

File: src/Symfony/Component/Uid/Uuid.php [hunk 2]
```diff
@@ -58,7 +58,7 @@ public static function fromString(string $uuid): parent
             $uuid = substr_replace($uuid, '-', 18, 0);
             $uuid = substr_replace($uuid, '-', 23, 0);
         } elseif (26 === \strlen($uuid) && Ulid::isValid($uuid)) {
-            $ulid = new Ulid('00000000000000000000000000');
+            $ulid = new NilUlid();
```
Comment (fancyweb) [inline src/Symfony/Component/Uid/Uuid.php:61]: Unrelated but let's leverage it.

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Component/Uid/Uuid.php [hunk 3]
```diff
@@ -117,6 +117,14 @@ final public static function v6(): UuidV6
 
     public static function isValid(string $uuid): bool
     {
+        if (self::NIL === $uuid && \in_array(static::class, [__CLASS__, NilUuid::class], true)) {
```
Comment (derrabus) [inline src/Symfony/Component/Uid/Uuid.php:120]: What is this `static::class` sorcery doing?

Comment (fancyweb) [reply src/Symfony/Component/Uid/Uuid.php:120]: `Uuid::isValid()` is used by all child classes. We want to allow `Uuid::isValid('0000...')` and `NilUuid::isValid('0000...')` but not `UuidV4::isValid('0000...')` for example.

Comment (derrabus) [reply src/Symfony/Component/Uid/Uuid.php:120]: Oh, I see. Thanks for the explanation. Wouldn't it make sense to override the method in those child classes instead? Either way, the change looks good to me.

**Review by derrabus — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @fancyweb.

File: src/Symfony/Component/Uid/Uuid.php (hunk 1 above)
Comment (fabpot) [reply src/Symfony/Component/Uid/Uuid.php:124]: cb97802e4f7e5dc8e49334d8159bd3d2411f92e5

**Pull request merged**